In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import missingno as msno

In [2]:
import os

print(os.getcwd())

c:\Users\Sachin Kuhikar\Documents\MedAssist-AI\notebooks


In [3]:
# Load all datasets

disease_df = pd.read_csv("../datasets/disease_prediction.csv")

diabetes_df = pd.read_csv("../datasets/dataset.csv")

patient_df = pd.read_csv("../datasets/Disease_symptom_and_patient_profile_dataset.csv")

description_df = pd.read_csv("../datasets/symptom_Description.csv")

precaution_df = pd.read_csv("../datasets/symptom_precaution.csv")

severity_df = pd.read_csv("../datasets/Symptom-severity.csv")

In [4]:
print("Disease Dataset:", disease_df.shape)
print("Diabetes Dataset:", diabetes_df.shape)
print("Patient Dataset:", patient_df.shape)
print("Description Dataset:", description_df.shape)
print("Precaution Dataset:", precaution_df.shape)
print("Severity Dataset:", severity_df.shape)

Disease Dataset: (42, 133)
Diabetes Dataset: (4920, 18)
Patient Dataset: (349, 10)
Description Dataset: (41, 2)
Precaution Dataset: (41, 5)
Severity Dataset: (133, 2)


In [5]:
datasets = {
    "Disease": disease_df,
    "Diabetes": diabetes_df,
    "Patient": patient_df,
    "Description": description_df,
    "Precaution": precaution_df,
    "Severity": severity_df
}

for name, df in datasets.items():
    print("="*60)
    print(f"{name} Dataset")
    print("="*60)
    print(df.head())
    print("\n")

Disease Dataset
   itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          0                     0                    1          1   
2        0          0                     0                    0          0   
3        1          0                     0                    0          0   
4        1          1                     0                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   
1       1           0             0        0                 0  ...   
2       0           0             1        1                 1  ...   
3       0           0             0        0                 0  ...   
4       0           0             1        0                 0  ...   

   blackheads  scurring  skin_peeling  silver_like_dusting  \
0           0       

In [6]:
for name, df in datasets.items():
    print("="*60)
    print(name)
    print(df.columns.tolist())
    print("\n")

Disease
['itching', 'skin_rash', 'nodal_skin_eruptions', 'continuous_sneezing', 'shivering', 'chills', 'joint_pain', 'stomach_pain', 'acidity', 'ulcers_on_tongue', 'muscle_wasting', 'vomiting', 'burning_micturition', 'spotting_ urination', 'fatigue', 'weight_gain', 'anxiety', 'cold_hands_and_feets', 'mood_swings', 'weight_loss', 'restlessness', 'lethargy', 'patches_in_throat', 'irregular_sugar_level', 'cough', 'high_fever', 'sunken_eyes', 'breathlessness', 'sweating', 'dehydration', 'indigestion', 'headache', 'yellowish_skin', 'dark_urine', 'nausea', 'loss_of_appetite', 'pain_behind_the_eyes', 'back_pain', 'constipation', 'abdominal_pain', 'diarrhoea', 'mild_fever', 'yellow_urine', 'yellowing_of_eyes', 'acute_liver_failure', 'fluid_overload', 'swelling_of_stomach', 'swelled_lymph_nodes', 'malaise', 'blurred_and_distorted_vision', 'phlegm', 'throat_irritation', 'redness_of_eyes', 'sinus_pressure', 'runny_nose', 'congestion', 'chest_pain', 'weakness_in_limbs', 'fast_heart_rate', 'pain_du

In [7]:
data2015 = pd.read_csv("../datasets/2015.csv")

print(data2015.shape)
print(data2015.columns.tolist())

data2015.head()

(441456, 330)
['_STATE', 'FMONTH', 'IDATE', 'IMONTH', 'IDAY', 'IYEAR', 'DISPCODE', 'SEQNO', '_PSU', 'CTELENUM', 'PVTRESD1', 'COLGHOUS', 'STATERES', 'CELLFON3', 'LADULT', 'NUMADULT', 'NUMMEN', 'NUMWOMEN', 'CTELNUM1', 'CELLFON2', 'CADULT', 'PVTRESD2', 'CCLGHOUS', 'CSTATE', 'LANDLINE', 'HHADULT', 'GENHLTH', 'PHYSHLTH', 'MENTHLTH', 'POORHLTH', 'HLTHPLN1', 'PERSDOC2', 'MEDCOST', 'CHECKUP1', 'BPHIGH4', 'BPMEDS', 'BLOODCHO', 'CHOLCHK', 'TOLDHI2', 'CVDINFR4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'ASTHNOW', 'CHCSCNCR', 'CHCOCNCR', 'CHCCOPD1', 'HAVARTH3', 'ADDEPEV2', 'CHCKIDNY', 'DIABETE3', 'DIABAGE2', 'SEX', 'MARITAL', 'EDUCA', 'RENTHOM1', 'NUMHHOL2', 'NUMPHON2', 'CPDEMO1', 'VETERAN3', 'EMPLOY1', 'CHILDREN', 'INCOME2', 'INTERNET', 'WEIGHT2', 'HEIGHT3', 'PREGNANT', 'QLACTLM2', 'USEEQUIP', 'BLIND', 'DECIDE', 'DIFFWALK', 'DIFFDRES', 'DIFFALON', 'SMOKE100', 'SMOKDAY2', 'STOPSMK2', 'LASTSMK2', 'USENOW3', 'ALCDAY5', 'AVEDRNK2', 'DRNK3GE5', 'MAXDRNKS', 'FRUITJU1', 'FRUIT1', 'FVBEANS', 'FVGREEN', 'FVORAN

,_STATE,FMONTH,IDATE,IMONTH,IDAY,IYEAR,DISPCODE,SEQNO,_PSU,CTELENUM,...,_PAREC1,_PASTAE1,_LMTACT1,_LMTWRK1,_LMTSCL1,_RFSEAT2,_RFSEAT3,_FLSHOT6,_PNEUMO2,_AIDTST3
0,1.0,1.0,b'01292015',b'01',b'29',b'2015',1200.0,2.015000e+09,2.015000e+09,1.0,...,4.0,2.0,1.0,1.0,1.0,1.0,1.0,NaN,NaN,1.0
1,1.0,1.0,b'01202015',b'01',b'20',b'2015',1100.0,2.015000e+09,2.015000e+09,1.0,...,2.0,2.0,3.0,3.0,4.0,2.0,2.0,NaN,NaN,2.0
2,1.0,1.0,b'02012015',b'02',b'01',b'2015',1200.0,2.015000e+09,2.015000e+09,1.0,...,9.0,9.0,9.0,9.0,9.0,9.0,9.0,9.0,9.0,NaN
3,1.0,1.0,b'01142015',b'01',b'14',b'2015',1100.0,2.015000e+09,2.015000e+09,1.0,...,4.0,2.0,1.0,1.0,1.0,1.0,1.0,NaN,NaN,9.0
4,1.0,1.0,b'01142015',b'01',b'14',b'2015',1100.0,2.015000e+09,2.015000e+09,1.0,...,4.0,2.0,1.0,1.0,1.0,1.0,1.0,NaN,NaN,1.0


In [8]:
# Dataset Summary

for name, df in datasets.items():
    print("="*70)
    print(f"{name.upper()} DATASET")
    print("="*70)

    print("Shape :", df.shape)
    print("Rows :", df.shape[0])
    print("Columns :", df.shape[1])

    print("\nFirst 5 Rows:")
    display(df.head())

DISEASE DATASET
Shape : (42, 133)
Rows : 42
Columns : 133

First 5 Rows:


,itching,skin_rash,nodal_skin_eruptions,continuous_sneezing,shivering,chills,joint_pain,stomach_pain,acidity,ulcers_on_tongue,...,blackheads,scurring,skin_peeling,silver_like_dusting,small_dents_in_nails,inflammatory_nails,blister,red_sore_around_nose,yellow_crust_ooze,prognosis
0,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection
1,0,0,0,1,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Allergy
2,0,0,0,0,0,0,0,1,1,1,...,0,0,0,0,0,0,0,0,0,GERD
3,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Chronic cholestasis
4,1,1,0,0,0,0,0,1,0,0,...,0,0,0,0,0,0,0,0,0,Drug Reaction


DIABETES DATASET
Shape : (4920, 18)
Rows : 4920
Columns : 18

First 5 Rows:


,Disease,Symptom_1,Symptom_2,Symptom_3,Symptom_4,Symptom_5,Symptom_6,Symptom_7,Symptom_8,Symptom_9,Symptom_10,Symptom_11,Symptom_12,Symptom_13,Symptom_14,Symptom_15,Symptom_16,Symptom_17
0,Fungal infection,itching,skin_rash,nodal_skin_eruptions,dischromic _patches,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Fungal infection,skin_rash,nodal_skin_eruptions,dischromic _patches,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Fungal infection,itching,nodal_skin_eruptions,dischromic _patches,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Fungal infection,itching,skin_rash,dischromic _patches,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Fungal infection,itching,skin_rash,nodal_skin_eruptions,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


PATIENT DATASET
Shape : (349, 10)
Rows : 349
Columns : 10

First 5 Rows:


,Disease,Fever,Cough,Fatigue,Difficulty Breathing,Age,Gender,Blood Pressure,Cholesterol Level,Outcome Variable
0,Influenza,Yes,No,Yes,Yes,19,Female,Low,Normal,Positive
1,Common Cold,No,Yes,Yes,No,25,Female,Normal,Normal,Negative
2,Eczema,No,Yes,Yes,No,25,Female,Normal,Normal,Negative
3,Asthma,Yes,Yes,No,Yes,25,Male,Normal,Normal,Positive
4,Asthma,Yes,Yes,No,Yes,25,Male,Normal,Normal,Positive


DESCRIPTION DATASET
Shape : (41, 2)
Rows : 41
Columns : 2

First 5 Rows:


,Disease,Description
0,Drug Reaction,An adverse drug reaction (ADR) is an injury ca...
1,Malaria,An infectious disease caused by protozoan para...
2,Allergy,An allergy is an immune system response to a f...
3,Hypothyroidism,"Hypothyroidism, also called underactive thyroi..."
4,Psoriasis,Psoriasis is a common skin disorder that forms...


PRECAUTION DATASET
Shape : (41, 5)
Rows : 41
Columns : 5

First 5 Rows:


,Disease,Precaution_1,Precaution_2,Precaution_3,Precaution_4
0,Drug Reaction,stop irritation,consult nearest hospital,stop taking drug,follow up
1,Malaria,Consult nearest hospital,avoid oily food,avoid non veg food,keep mosquitos out
2,Allergy,apply calamine,cover area with bandage,NaN,use ice to compress itching
3,Hypothyroidism,reduce stress,exercise,eat healthy,get proper sleep
4,Psoriasis,wash hands with warm soapy water,stop bleeding using pressure,consult doctor,salt baths


SEVERITY DATASET
Shape : (133, 2)
Rows : 133
Columns : 2

First 5 Rows:


,Symptom,weight
0,itching,1
1,skin_rash,3
2,nodal_skin_eruptions,4
3,continuous_sneezing,4
4,shivering,5


In [9]:
for name, df in datasets.items():
    print("="*70)
    print(name)
    print(df.dtypes)

Disease
itching                 int64
skin_rash               int64
nodal_skin_eruptions    int64
continuous_sneezing     int64
shivering               int64
                        ...  
inflammatory_nails      int64
blister                 int64
red_sore_around_nose    int64
yellow_crust_ooze       int64
prognosis                 str
Length: 133, dtype: object
Diabetes
Disease       str
Symptom_1     str
Symptom_2     str
Symptom_3     str
Symptom_4     str
Symptom_5     str
Symptom_6     str
Symptom_7     str
Symptom_8     str
Symptom_9     str
Symptom_10    str
Symptom_11    str
Symptom_12    str
Symptom_13    str
Symptom_14    str
Symptom_15    str
Symptom_16    str
Symptom_17    str
dtype: object
Patient
Disease                   str
Fever                     str
Cough                     str
Fatigue                   str
Difficulty Breathing      str
Age                     int64
Gender                    str
Blood Pressure            str
Cholesterol Level         str
Outcome Va

In [10]:
for name, df in datasets.items():
    print("="*70)
    print(name)

    missing = df.isnull().sum()

    print(missing[missing > 0])

    print()

Disease
Series([], dtype: int64)

Diabetes
Symptom_4      348
Symptom_5     1206
Symptom_6     1986
Symptom_7     2652
Symptom_8     2976
Symptom_9     3228
Symptom_10    3408
Symptom_11    3726
Symptom_12    4176
Symptom_13    4416
Symptom_14    4614
Symptom_15    4680
Symptom_16    4728
Symptom_17    4848
dtype: int64

Patient
Series([], dtype: int64)

Description
Series([], dtype: int64)

Precaution
Precaution_3    1
Precaution_4    1
dtype: int64

Severity
Series([], dtype: int64)



In [11]:
for name, df in datasets.items():
    print("="*70)
    print(name)
    print("Duplicate Rows :", df.duplicated().sum())

Disease
Duplicate Rows : 0
Diabetes
Duplicate Rows : 4616
Patient
Duplicate Rows : 49
Description
Duplicate Rows : 0
Precaution
Duplicate Rows : 0
Severity
Duplicate Rows : 0


In [12]:
for name, df in datasets.items():
    print("="*70)
    print(name)
    display(df.describe(include='all'))

Disease


,itching,skin_rash,nodal_skin_eruptions,continuous_sneezing,shivering,chills,joint_pain,stomach_pain,acidity,ulcers_on_tongue,...,blackheads,scurring,skin_peeling,silver_like_dusting,small_dents_in_nails,inflammatory_nails,blister,red_sore_around_nose,yellow_crust_ooze,prognosis
count,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,...,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42.000000,42
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,41
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Fungal infection
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
mean,0.166667,0.190476,0.023810,0.047619,0.023810,0.166667,0.142857,0.047619,0.047619,0.023810,...,0.023810,0.023810,0.047619,0.023810,0.023810,0.023810,0.023810,0.047619,0.023810,NaN
std,0.377195,0.397437,0.154303,0.215540,0.154303,0.377195,0.354169,0.215540,0.215540,0.154303,...,0.154303,0.154303,0.215540,0.154303,0.154303,0.154303,0.154303,0.215540,0.154303,NaN
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN
25%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN
50%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN
75%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN


Diabetes


,Disease,Symptom_1,Symptom_2,Symptom_3,Symptom_4,Symptom_5,Symptom_6,Symptom_7,Symptom_8,Symptom_9,Symptom_10,Symptom_11,Symptom_12,Symptom_13,Symptom_14,Symptom_15,Symptom_16,Symptom_17
count,4920,4920,4920,4920,4572,3714,2934,2268,1944,1692,1512,1194,744,504,306,240,192,72
unique,41,34,48,54,50,38,32,26,21,22,21,18,11,8,4,3,3,1
top,Fungal infection,vomiting,vomiting,fatigue,high_fever,headache,nausea,abdominal_pain,abdominal_pain,yellowing_of_eyes,yellowing_of_eyes,irritability,malaise,muscle_pain,chest_pain,chest_pain,blood_in_sputum,muscle_pain
freq,120,822,870,726,378,348,390,264,276,228,198,120,126,72,96,144,72,72


Patient


,Disease,Fever,Cough,Fatigue,Difficulty Breathing,Age,Gender,Blood Pressure,Cholesterol Level,Outcome Variable
count,349,349,349,349,349,349.000000,349,349,349,349
unique,116,2,2,2,2,NaN,2,3,3,2
top,Asthma,Yes,No,Yes,No,NaN,Female,High,High,Positive
freq,23,175,182,242,261,NaN,176,167,166,186
mean,NaN,NaN,NaN,NaN,NaN,46.323782,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,13.085090,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,19.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,35.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,45.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,55.000000,NaN,NaN,NaN,NaN


Description


,Disease,Description
count,41,41
unique,41,41
top,Drug Reaction,An adverse drug reaction (ADR) is an injury ca...
freq,1,1


Precaution


,Disease,Precaution_1,Precaution_2,Precaution_3,Precaution_4
count,41,41,41,40,40
unique,41,32,34,30,24
top,Drug Reaction,Consult nearest hospital,exercise,consult doctor,follow up
freq,1,3,3,6,6


Severity


,Symptom,weight
count,133,133.000000
unique,132,NaN
top,fluid_overload,NaN
freq,2,NaN
mean,NaN,4.225564
std,NaN,1.323543
min,NaN,1.000000
25%,NaN,3.000000
50%,NaN,4.000000
75%,NaN,5.000000


# Observations

## Disease Dataset

- 42 diseases
- 132 symptom columns
- Target column: prognosis

## Diabetes Dataset

- Disease with symptom list
- Symptoms stored as text

## Patient Dataset

- Contains demographic information
- Age, Gender, Blood Pressure, Cholesterol

## Description Dataset

- Maps disease to description

## Precaution Dataset

- Maps disease to precautions

## Severity Dataset

- Maps symptoms to severity weights

In [13]:
diabetes_df = diabetes_df.fillna("None")

In [14]:
diabetes_df.isnull().sum()

Disease       0
Symptom_1     0
Symptom_2     0
Symptom_3     0
Symptom_4     0
Symptom_5     0
Symptom_6     0
Symptom_7     0
Symptom_8     0
Symptom_9     0
Symptom_10    0
Symptom_11    0
Symptom_12    0
Symptom_13    0
Symptom_14    0
Symptom_15    0
Symptom_16    0
Symptom_17    0
dtype: int64

In [15]:
patient_df = patient_df.drop_duplicates()

In [16]:
patient_df.shape

(300, 10)

In [17]:
print("Original Shape:", diabetes_df.shape)
print("After Removing Duplicates:", diabetes_df.drop_duplicates().shape)

Original Shape: (4920, 18)
After Removing Duplicates: (304, 18)


In [18]:
df_diabetes.to_csv("../data/processed/diabetes_clean.csv", index=False)

print("Saved Successfully!")

NameError: name 'df_diabetes' is not defined

In [ ]:
%whos

Variable         Type         Data/Info
---------------------------------------
data2015         DataFrame    Shape: (441456, 330)
datasets         dict         n=6
description_df   DataFrame    Shape: (41, 2)
df               DataFrame    Shape: (133, 2)
diabetes_df      DataFrame    Shape: (4920, 18)
disease_df       DataFrame    Shape: (42, 133)
missing          Series       Shape: (2,)
msno             module       <module 'missingno' from <...>\missingno\\__init__.py'>
name             str          Severity
np               module       <module 'numpy' from 'c:\<...>ges\\numpy\\__init__.py'>
os               module       <module 'os' (frozen)>
patient_df       DataFrame    Shape: (300, 10)
pd               module       <module 'pandas' from 'c:<...>es\\pandas\\__init__.py'>
plt              module       <module 'matplotlib.pyplo<...>\\matplotlib\\pyplot.py'>
precaution_df    DataFrame    Shape: (41, 5)
severity_df      DataFrame    Shape: (133, 2)
sns              module       <mo

In [ ]:
diabetes_df = diabetes_df.drop_duplicates()

print(diabetes_df.shape)

(304, 18)


In [ ]:
diabetes_df.to_csv("../processed_data/diabetes_clean.csv", index=False)

print("Saved Successfully!")

Saved Successfully!


In [ ]:
diabetes_df.to_csv("../processed_data/diabetes_clean.csv", index=False)

print("✅ Diabetes dataset saved successfully!")

✅ Diabetes dataset saved successfully!


In [ ]:
patient_df.to_csv("../processed_data/patient_clean.csv", index=False)

print("✅ Patient dataset saved successfully!")

✅ Patient dataset saved successfully!


In [ ]:
disease_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 42 entries, 0 to 41
Columns: 133 entries, itching to prognosis
dtypes: int64(132), str(1)
memory usage: 43.8 KB


In [ ]:
disease_df.to_csv("../processed_data/disease_clean.csv", index=False)

print("✅ Disease dataset saved successfully!")

✅ Disease dataset saved successfully!


In [ ]:
print("Disease Dataset Diseases:")
print(sorted(disease_df["prognosis"].unique()))

print("\n" + "="*80 + "\n")

print("Diabetes Dataset Diseases:")
print(sorted(diabetes_df["Disease"].unique()))

print("\n" + "="*80 + "\n")

print("Patient Dataset Diseases:")
print(sorted(patient_df["Disease"].unique()))

Disease Dataset Diseases:
['(vertigo) Paroymsal  Positional Vertigo', 'AIDS', 'Acne', 'Alcoholic hepatitis', 'Allergy', 'Arthritis', 'Bronchial Asthma', 'Cervical spondylosis', 'Chicken pox', 'Chronic cholestasis', 'Common Cold', 'Dengue', 'Diabetes ', 'Dimorphic hemmorhoids(piles)', 'Drug Reaction', 'Fungal infection', 'GERD', 'Gastroenteritis', 'Heart attack', 'Hepatitis B', 'Hepatitis C', 'Hepatitis D', 'Hepatitis E', 'Hypertension ', 'Hyperthyroidism', 'Hypoglycemia', 'Hypothyroidism', 'Impetigo', 'Jaundice', 'Malaria', 'Migraine', 'Osteoarthristis', 'Paralysis (brain hemorrhage)', 'Peptic ulcer diseae', 'Pneumonia', 'Psoriasis', 'Tuberculosis', 'Typhoid', 'Urinary tract infection', 'Varicose veins', 'hepatitis A']


Diabetes Dataset Diseases:
['(vertigo) Paroymsal  Positional Vertigo', 'AIDS', 'Acne', 'Alcoholic hepatitis', 'Allergy', 'Arthritis', 'Bronchial Asthma', 'Cervical spondylosis', 'Chicken pox', 'Chronic cholestasis', 'Common Cold', 'Dengue', 'Diabetes ', 'Dimorphic hemm

In [ ]:
for col in diabetes_df.columns[1:]:
    print(col)

Symptom_1
Symptom_2
Symptom_3
Symptom_4
Symptom_5
Symptom_6
Symptom_7
Symptom_8
Symptom_9
Symptom_10
Symptom_11
Symptom_12
Symptom_13
Symptom_14
Symptom_15
Symptom_16
Symptom_17


In [ ]:
display(diabetes_df.head())

,Disease,Symptom_1,Symptom_2,Symptom_3,Symptom_4,Symptom_5,Symptom_6,Symptom_7,Symptom_8,Symptom_9,Symptom_10,Symptom_11,Symptom_12,Symptom_13,Symptom_14,Symptom_15,Symptom_16,Symptom_17
0,Fungal infection,itching,skin_rash,nodal_skin_eruptions,dischromic _patches,None,None,None,None,None,None,None,None,None,None,None,None,None
1,Fungal infection,skin_rash,nodal_skin_eruptions,dischromic _patches,None,None,None,None,None,None,None,None,None,None,None,None,None,None
2,Fungal infection,itching,nodal_skin_eruptions,dischromic _patches,None,None,None,None,None,None,None,None,None,None,None,None,None,None
3,Fungal infection,itching,skin_rash,dischromic _patches,None,None,None,None,None,None,None,None,None,None,None,None,None,None
4,Fungal infection,itching,skin_rash,nodal_skin_eruptions,None,None,None,None,None,None,None,None,None,None,None,None,None,None


In [ ]:
# Create an empty DataFrame with the same columns as disease_df
diabetes_transformed = pd.DataFrame(columns=disease_df.columns)

print(diabetes_transformed.shape)

(0, 133)


In [ ]:
# Take the first row from diabetes_df
row = diabetes_df.iloc[0]

print(row)

Disease            Fungal infection
Symptom_1                   itching
Symptom_2                 skin_rash
Symptom_3      nodal_skin_eruptions
Symptom_4       dischromic _patches
Symptom_5                      None
Symptom_6                      None
Symptom_7                      None
Symptom_8                      None
Symptom_9                      None
Symptom_10                     None
Symptom_11                     None
Symptom_12                     None
Symptom_13                     None
Symptom_14                     None
Symptom_15                     None
Symptom_16                     None
Symptom_17                     None
Name: 0, dtype: str


In [ ]:
# Create one empty row with object data type
new_row = pd.Series(index=disease_df.columns, dtype="object")

# Set all symptom columns to 0
new_row.iloc[:-1] = 0

# Set prognosis
new_row["prognosis"] = row["Disease"]

print(new_row.head())

itching                 0
skin_rash               0
nodal_skin_eruptions    0
continuous_sneezing     0
shivering               0
dtype: object


In [ ]:
# Go through all symptom columns in this patient
for symptom in row[1:]:   # Skip the Disease column

    if symptom != "None":
        new_row[symptom] = 1

# Show only the symptoms that are 1
print(new_row[new_row == 1])

print("\nDisease:")
print(new_row["prognosis"])

itching                  1
 skin_rash               1
 nodal_skin_eruptions    1
 dischromic _patches     1
dtype: object

Disease:
Fungal infection


In [ ]:
# Empty list to store transformed rows
transformed_rows = []

# Loop through every row in diabetes_df
for _, row in diabetes_df.iterrows():

    # Create an empty row
    new_row = pd.Series(index=disease_df.columns, dtype="object")
    new_row.iloc[:-1] = 0

    # Set disease name
    new_row["prognosis"] = row["Disease"]

    # Mark symptoms as 1
    for symptom in row[1:]:
        if symptom != "None":
            new_row[symptom] = 1

    # Add row to list
    transformed_rows.append(new_row)

# Convert list into DataFrame
diabetes_transformed = pd.DataFrame(transformed_rows)

print(diabetes_transformed.shape)

(304, 263)


In [ ]:
# Columns that exist in diabetes_transformed but not in disease_df
extra_cols = set(diabetes_transformed.columns) - set(disease_df.columns)

print("Number of extra columns:", len(extra_cols))
print(sorted(extra_cols))

Number of extra columns: 130
[' abdominal_pain', ' abnormal_menstruation', ' acidity', ' acute_liver_failure', ' altered_sensorium', ' anxiety', ' back_pain', ' belly_pain', ' blackheads', ' bladder_discomfort', ' blister', ' blood_in_sputum', ' bloody_stool', ' blurred_and_distorted_vision', ' breathlessness', ' brittle_nails', ' bruising', ' burning_micturition', ' chest_pain', ' chills', ' cold_hands_and_feets', ' coma', ' congestion', ' constipation', ' continuous_feel_of_urine', ' continuous_sneezing', ' cough', ' cramps', ' dark_urine', ' dehydration', ' depression', ' diarrhoea', ' dischromic _patches', ' distention_of_abdomen', ' dizziness', ' drying_and_tingling_lips', ' enlarged_thyroid', ' excessive_hunger', ' extra_marital_contacts', ' family_history', ' fast_heart_rate', ' fatigue', ' fluid_overload', ' foul_smell_of urine', ' headache', ' high_fever', ' hip_joint_pain', ' history_of_alcohol_consumption', ' increased_appetite', ' indigestion', ' inflammatory_nails', ' inte

In [ ]:
# Empty list to store transformed rows
transformed_rows = []

# Loop through every row
for _, row in diabetes_df.iterrows():

    # Create empty row
    new_row = pd.Series(index=disease_df.columns, dtype="object")
    new_row.iloc[:-1] = 0

    # Disease name
    new_row["prognosis"] = row["Disease"].strip()

    # Symptoms
    for symptom in row[1:]:

        if pd.notna(symptom) and symptom != "None":

            symptom = symptom.strip()   # ⭐ Remove leading/trailing spaces

            if symptom in disease_df.columns:
                new_row[symptom] = 1

    transformed_rows.append(new_row)

# Create DataFrame
diabetes_transformed = pd.DataFrame(transformed_rows)

print(diabetes_transformed.shape)

(304, 133)


In [ ]:
diabetes_transformed.to_csv(
    "../processed_data/diabetes_transformed.csv",
    index=False
)

print("✅ Diabetes transformed dataset saved successfully!")

✅ Diabetes transformed dataset saved successfully!


In [19]:
# Load processed datasets

disease_df = pd.read_csv("../processed_data/disease_clean.csv")
diabetes_df = pd.read_csv("../processed_data/diabetes_transformed.csv")
patient_df = pd.read_csv("../processed_data/patient_clean.csv")

In [20]:
print("Disease:", disease_df.shape)
print("Diabetes:", diabetes_df.shape)
print("Patient:", patient_df.shape)

Disease: (42, 133)
Diabetes: (304, 133)
Patient: (300, 10)


In [21]:
def clean_disease_name(name):
    if pd.isna(name):
        return name

    name = str(name).strip().lower()

    replacements = {
        "diabetes ": "diabetes",
        "hypertension ": "hypertension",
        "hepatitis a": "hepatitis a",
        "chicken pox": "chickenpox",
        "urinary tract infection": "urinary tract infection",
        "typhoid": "typhoid fever",
        "heart attack": "myocardial infarction (heart attack)",
        "bronchial asthma": "asthma",
        "common cold": "common cold",
        "osteoarthristis": "osteoarthritis"
    }

    return replacements.get(name, name)

In [22]:
disease_df["prognosis"] = disease_df["prognosis"].apply(clean_disease_name)

diabetes_df["prognosis"] = diabetes_df["prognosis"].apply(clean_disease_name)

patient_df["Disease"] = patient_df["Disease"].apply(clean_disease_name)

print("Done!")

Done!


In [23]:
common = set(disease_df["prognosis"]) & set(diabetes_df["prognosis"]) & set(patient_df["Disease"])

print("Common diseases:", len(common))
print(sorted(common))

Common diseases: 19
['acne', 'asthma', 'chickenpox', 'common cold', 'diabetes', 'gastroenteritis', 'hepatitis b', 'hypertension', 'hyperthyroidism', 'hypoglycemia', 'hypothyroidism', 'malaria', 'migraine', 'osteoarthritis', 'pneumonia', 'psoriasis', 'tuberculosis', 'typhoid fever', 'urinary tract infection']


In [24]:
# Merge both datasets

final_dataset = pd.concat(
    [disease_df, diabetes_df],
    ignore_index=True
)

print("Shape before removing duplicates:", final_dataset.shape)

Shape before removing duplicates: (346, 133)


In [25]:
duplicates = final_dataset.duplicated().sum()

print("Duplicate rows:", duplicates)

Duplicate rows: 40


In [26]:
# Remove duplicate rows

final_dataset = final_dataset.drop_duplicates()

print("Final Shape:", final_dataset.shape)

Final Shape: (306, 133)


In [27]:
final_dataset.to_csv(
    "../processed_data/final_dataset.csv",
    index=False
)

print("✅ Final dataset saved successfully!")

✅ Final dataset saved successfully!
